sunkit_image/granule.py

```diff
@@ -0,0 +1,331 @@
+        1 = in-between.
+    """
+
+    if llambda_axis not in [-1, 2]:
+        raise Exception(
+            "Wrong data shape. \
+        (either scalar or (x, y, llambda) )"
+        )
+    n_clusters = 3
+    n_init = 20
+    x_size = np.shape(data)[0]
+    y_size = np.shape(data)[1]
+    if llambda_axis == -1:
+        data_flat = np.reshape(data, (x_size * y_size, 1))
+        labels_flat = KMeans(n_clusters, n_init).fit(data_flat).labels_
+        labels = np.reshape(labels_flat, (x_size, y_size))
+    else:
+        llambda_size = np.shape(data)[llambda_axis]
+        data = np.reshape(data, (x_size * y_size, llambda_size))
+        labels = np.reshape(KMeans(n_clusters, n_init).fit(data), (x_size, y_size))
+
+    # make intergranules 0, granules 1:
+    group0_mean = np.mean(data[labels == 0])
+    group1_mean = np.mean(data[labels == 1])
+    group2_mean = np.mean(data[labels == 2])
+
+    # intergranules
+    min_index = np.argmin([group0_mean, group1_mean, group2_mean])
+    segmented_map = np.ones(labels.shape)
+
+    segmented_map[[labels[:, :] == min_index][0]] -= 1
+
+    return segmented_map
+
+
+def cross_correlation(segment1, segment2):
+    """
+    Return -1 and logging.info a message if the agreement between two arrays is
+    low, 0 otherwise. Designed to be used with segment and segment_kmeans
+    function.
+
+    Parameters
+    ----------
+    segment1 : `numpy.ndarray`
+        'main' array to compare the other input array against.
+    segment2 : `numpy.ndarray`
+        'other' array (i.e. data segmented using kmeans).
+
+    Returns
+    -------
+    [label, confidence] : `list`
+        'Label' (int) summarizes the confidence metric -
+            -1: if agreement is low (below 75%)
+            0: otherwise
+        'confidence' (float) is the numeric confidence metric -
+            float between 0 and 1 (0 if no agreement, 1 if completely agrees).
+    """
+
+    total_granules = np.count_nonzero(segment1 == 1)
+    total_intergranules = np.count_nonzero(segment1 == 0)
+
+    if total_granules == 0:
+        raise Exception("clustering problematic (no granules found)")
+
+    if total_intergranules == 0:
+        raise Exception("clustering problematic (no intergranules found)")
+
+    x_size = np.shape(segment1)[0]
+    y_size = np.shape(segment1)[1]
+
+    granule_agreement_count = 0
+    intergranule_agreement_count = 0
+    for i in range(x_size):
+        for j in range(y_size):
+            if segment1[i, j] == 1 and segment2[i, j] == 1:
+                granule_agreement_count += 1
+            elif segment1[i, j] == 0 and segment2[i, j] == 0:
+                intergranule_agreement_count += 1
+
+    percentage_agreement_granules = granule_agreement_count / total_granules
+    percentage_agreement_intergranules = intergranule_agreement_count / total_intergranules
+    try:
+        confidence = np.mean([percentage_agreement_granules, percentage_agreement_intergranules])
+    except TypeError:

```

Comment:
What causes the TypeError?

Reply:
Hmm... that is a good point. I can't think of any reason that taking either `percentage_agreement_granules` or `percentage_agreement_intergranules` would not be int or float. I will ask why this was put there, but I think I can just get rid of it.

Reply:
Actually, I think this might be good to have, as it will occur if either `percentage_agreement_granules` or `percentage_agreement_intergranules` are NaN, which would occur if there was something wrong with the data and `total_granules` or `total_intergranules` were zero.

Reply:
If there are nans in either of those, would it not be better if the code just errored?

Reply:
That would certainly give the user more information. I have been advised to catch all python errors in exceptions, but is this a place it might make sense to simply let the error occur? Otherwise I suppose I could throw a divide by zero error myself.

Reply:
My gut feeling is to let the error happen, is there an upside to letting the code continue with 0 confidence?